### PR TITLE
Inset Mac OS traffic lights

### PR DESCRIPTION
**Pull request by haasal**
### What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Docs
- [x] New Binding issue tauri-apps/tauri#4789
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

### Does this PR introduce a breaking change?

- [ ] Yes, and the changes were approved in issue #___
- [x] No

### Checklist
- [x] When resolving issues, they are referenced in the PR's title (e.g `fix: remove a typo, closes #___, #___`)
- [ ] A change file is added if any packages will require a version bump due to this PR per [the instructions in the readme](https://github.com/tauri-apps/tauri/blob/dev/.changes/readme.md).
- [x] I have added a convincing reason for adding this feature, if necessary

### Other information

**Important**: There still exist artifacts when resizing the window because the traffic lights are set just after the window has already been drawn.

### CODE REVIEW

**Comment by morajabi** [comment]
Thanks for this awesome PR! I had tried doing this only to fail at the "resize" artifacts: #494 Reading your solution made it clear. I should close that issue then.

**Comment by PhotonQuantum** [comment]
This feature is fantastic!

However, I noticed a strange behavior: the hover effect still shows when I move the cursor to the "original" position of traffic lights. The electron implementation doesn't seem like have this behavior. I dug into its code but failed to find a solution.

Update: strangely, the problem goes away if I resize the window.

**Comment by tr3ysmith** [comment]
Hey what's the status of this? I really would like to be able to apply inset on the traffic lights without work arounds

**Comment by haasal** [comment]
Hi, me too. But I know really little about Tao. If this gets merged then we still have to implement it somehow into tauri.
Could someone review this please?

**Comment by wusyong** [comment]
The implementation seems fine to me. But I'm not sure if it's good to merge now.
@amrbashir Do you think winit can accept this?
Btw you can get the raw window handle to change it yourself if you can't wait for it any longer.

**Comment by amrbashir** [comment]
there is a good chance they may not accept but lets go ahead and merge this and if they didn't accept it upstream, we can implement it directly in tauri

File: examples/macos_traffic_lights.rs
Comment (amrbashir) [inline examples/macos_traffic_lights.rs]: example not needed

File: src/platform/macos.rs
Comment (amrbashir) [inline src/platform/macos.rs]: missing documentation and use `LogicalSize` instead of a tuple

File: src/platform/macos.rs
Comment (amrbashir) [inline src/platform/macos.rs]: missing documentation and use `LogicalSize` instead of a tuple

File: src/platform_impl/macos/window.rs
Comment (amrbashir) [inline src/platform_impl/macos/window.rs]: why modify this?

File: src/platform_impl/macos/window.rs
Comment (amrbashir) [inline src/platform_impl/macos/window.rs]:
```suggestion
```

File: src/platform_impl/macos/window.rs
Comment (amrbashir) [inline src/platform_impl/macos/window.rs]:
```suggestion
unsafe fn position_traffic_lights<W: NSWindow + Copy>(window: W, x: f64, y: f64) {
```

File: src/platform_impl/macos/window.rs
Comment (amrbashir) [inline src/platform_impl/macos/window.rs]:
```suggestion
  let close_rect = close.frame();
```

File: src/platform_impl/macos/window.rs
Comment (amrbashir) [inline src/platform_impl/macos/window.rs]:
```suggestion
  let mut title_bar_rect = title_bar_container_view.frame();
```

File: src/platform_impl/macos/window.rs
Comment (amrbashir) [inline src/platform_impl/macos/window.rs]:
```suggestion
  title_bar_rect.origin.y = window.frame().size.height - title_bar_frame_height;
```

File: src/platform_impl/macos/window.rs
Comment (amrbashir) [inline src/platform_impl/macos/window.rs]:
```suggestion
  let space_between = miniaturize.frame().origin.x - close_rect.origin.x;
```

File: src/platform_impl/macos/window.rs
Comment (amrbashir) [inline src/platform_impl/macos/window.rs]:
```suggestion
  let title_bar_frame_height = close_rect.size.height + y;
```

File: src/platform_impl/macos/window.rs
Comment (amrbashir) [inline src/platform_impl/macos/window.rs]:
```suggestion
    let mut rect = button.frame();
```

**Review by amrbashir — CHANGES_REQUESTED** [review]

File: src/platform_impl/macos/window.rs
Comment (haasal) [reply src/platform_impl/macos/window.rs]: Honestly... I have no clue. I worked on this a long time ago

**Comment by haasal** [comment]
I'm sorry this has taken so long. I didn't really have time to work on this. I will document the function and delete the example these days. I hope it's not to late to merge?

**Review by amrbashir — COMMENTED** [review]
Looks fine to me, however I'll hand this to @wusyong and @JonasKruckenberg (since you were interested in this)

**Comment by amrbashir** [comment]
Please rebase the PR to fix the conflicts and add a change file

**Comment by JonasKruckenberg** [comment]
I'll review this tomorrow, it's a super important imo and this or has been open wayy too long 😉 let's get this over the finish line

**Comment by haasal** [comment]
Maybe it's important to add that I couldn't test the changes since my MacBook broke 🥺. If one of you could do that it would be amazing

**Comment by JonasKruckenberg** [comment]
After testing this PR now, I feel `LogicalSize` makes no sense here, the way you currently call the method is as such

```rust
                                      // WTH means width here? and height?
.with_traffic_light_inset(LogicalSize { width: 10, height: 20 });
```

**Comment by JonasKruckenberg** [comment]
Okay a few other observations:

1. While this does successfully inset the traffic lights, it does not increase the draggable area size
2. When resizing the window, it looses the inset

The second one really is a dealbreaker right now, I guess we need to implement a delegate and stuff just like electron does as well

**Comment by haasal** [comment]
> Okay a few other observations:
> 
>     1. While this does successfully inset the traffic lights, it does not increase the draggable area size
> 
>     2. When resizing the window, it looses the inset
> 
> 
> The second one really is a dealbreaker right now, I guess we need to implement a delegate and stuff just like electron does as well

I don't know if you fixed it already in that commit, but when I tested it with my supplied example the inset was persistent during resizing of the window.

**Comment by JonasKruckenberg** [comment]
> I don't know if you fixed it already in that commit, but when I tested it with my supplied example the inset was persistent during resizing of the window.

Okay yes my bad, calling the `set_traffic_light_inset` method after the RedrawRequested event got fired fixes the artifacts.

My only question @haasal is why the `with_traffic_light_inset` method exists if it's not really doing what the user expects. This feels like misleading api design

**Comment by JonasKruckenberg** [comment]
@amrbashir I added back the example, since it's indeed needed to use the method correctly (it's not enough to just call it once but it needs to be called after `RedrawRequested`)

**Comment by haasal** [comment]
Yeah that was the reason I made the example. I already tried implementing this in Tauri itself using the raw window handle but I couldn't get rid of these artifacts. Do you think this method can now be used in tauri without causing artifacts?

**Comment by haasal** [comment]
> My only question @haasal is why the `with_traffic_light_inset` method exists if it's not really doing what the user expects. This feels like misleading api design

I don't know how to implement it instead. This whole thing gave me big headaches because I couldn't find a way to set the inset without having to call the method in some weird way. If you have any idea how to do this without having to call the method after `RedrawRequested` it would be great!

**Comment by amrbashir** [comment]
I still don't see why the example is needed since tao should already handle that internally and not require users to keep calling it in `RedrawRequested`

**Comment by JonasKruckenberg** [comment]
> I still don't see why the example is needed since tao should already handle that internally and not require users to keep calling it in `RedrawRequested`

then we should change it to behave that way

**Comment by haasal** [comment]
I really would like to get this over with but as said earlier I don't really know how. If you know how to directly inject something into the "window draw pipeline" (I.e. how to automatically call my inset-method/function without requiring the user to do some weird setup) I could take a look at it

**Comment by amrbashir** [comment]
@haasal you need add an inset field on the `ViewState` struct here https://github.com/tauri-apps/tao/blob/c5d606dffeb1733ab06fd8c43eb3b9e7b2f553fe/src/platform_impl/macos/view.rs#L54 and your new function `set_traffic_light_inset` sets this field, then in `draw_rect` https://github.com/tauri-apps/tao/blob/c5d606dffeb1733ab06fd8c43eb3b9e7b2f553fe/src/platform_impl/macos/view.rs#L382  you have access to the state where you can read it and re-position the tarffic lights

**Comment by haasal** [comment]
But the `with_traffic_light_inset()` method should remain as the public API method for the user to call, right?
The user just shouldn't ever have to call `set_traffic_light_inset()` manually.

**Comment by amrbashir** [comment]
@haasal No, both methods should be public, both should modify the newly added state and triggers the traffic light positioning

**Comment by haasal** [comment]
Sorry for not contributing to this in the last weeks. In spite of what I said I noticed that I don't currently have the time to finish this. I would really appreciate if one of you could push this over the finish line.

**Comment by haasal** [comment]
Got a new mac and fixed it. I removed the set_traffic_light_inset() method because it wasn't really necessary anymore and would add unnecessary complexity. The with_traffic_light_inset() method on WindowBuilder should now work as expected.

Edit: I added the `set_traffic_light_inset()` back in. I overestimated how complicated it would be to implement this

**Comment by haasal** [comment]
Imho apart from minor refactoring (if you see required) this should be ready to merge. Could you maybe take a look at this @JonasKruckenberg @amrbashir?

File: src/platform_impl/macos/view.rs
Comment (amrbashir) [inline src/platform_impl/macos/view.rs]: why was this check removed?

File: src/platform_impl/macos/window.rs
Comment (amrbashir) [inline src/platform_impl/macos/window.rs]:
```suggestion
```

File: src/platform_impl/macos/window.rs
Comment (amrbashir) [inline src/platform_impl/macos/window.rs]:
```suggestion
  fn set_traffic_light_position<P: Into<Position>>(&self, position: P) {
```

**Review by amrbashir — CHANGES_REQUESTED** [review]

File: src/platform_impl/macos/view.rs
Comment (haasal) [reply src/platform_impl/macos/view.rs]: Must have been on accident when deleting my own stuff

File: src/platform_impl/macos/window.rs
Comment (haasal) [reply src/platform_impl/macos/window.rs]: I don't really understand this. Do you want me to change only the public interface (i.e. set_traffic_..., with_...) or should I make the WindowBuilder field and the View attributes (or how they are called) also into a `position`? How do I extract x and y from this?

The x and y for the inset are LogicalPosition coordinates right? Can't we just use LogicalPosition as the inset Type instead of Position? This can be called passed using `LogicalPosition {x, y}` and so Jonas' argument against `LogicalSize` is satisfied

**Comment by haasal** [comment]
@amrbashir ok, I implemented your feedback. I hope the `LogicalPosition` is ok

File: src/platform_impl/macos/window.rs
Comment (amrbashir) [reply src/platform_impl/macos/window.rs]: This change should be added to all public APIs, and then you can convert them into logical position internally before applying the inset.

Comment (amrbashir) [reply src/platform_impl/macos/window.rs]: Taking `Into<Position>` instead of `Into<LogicalPosition>` allows more flexibility and also to be consistent with other public APIs.

Comment (haasal) [reply src/platform_impl/macos/window.rs]: The issue is that in the `set_traffic_light_position()` implementation I don't have access to the `self.window.scale_factor()` method for `WindowBuilder` which is needed to convert from `Position` to `LogicalPosition`

Comment (amrbashir) [reply src/platform_impl/macos/window.rs]: you can get the scale factor using `NSWindow::backingScaleFactor(ns_window)` like this https://github.com/tauri-apps/tao/blob/bd8bafeca8a063351022cd5d5791f3111e8761d0/src/platform_impl/macos/window.rs#L820

Comment (haasal) [reply src/platform_impl/macos/window.rs]: Resolved?

**Review by amrbashir — APPROVED** [review]

**Comment by amrbashir** [comment]
Thank you

**Comment by amrbashir** [comment]
Just waiting for @wusyong to test and approve as well.

**Comment by amrbashir** [comment]
@haasal looks like you didn't sign your commits and I can't merge the PR unless they are signed. You need to setup [commit signing](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits), then you can sign past commit like [this](https://superuser.com/questions/397149/can-you-gpg-sign-old-commits/1123928#1123928) for example.

**Comment by haasal** [comment]
@amrbashir Do i also need to sign the partially verified commits? Also @JonasKruckenberg 's commits seem not to be signed either, right?

**Comment by wusyong** [comment]
> Just waiting for @wusyong to test and approve as well.

I'm okay with the API but the title bar is expanded. Is this what you expect?

Screenshot 2023-10-26 at 6 41 42 PM

**Comment by haasal** [comment]
Oh god... I messed up didn't I?
@wusyong yes this is what one should expect. This will only work (look) properly with the overlay titlebar type

**Comment by wusyong** [comment]
You can squash all your commits into one and then merge with dev. This should be the easiest way to sign commit.

**Comment by haasal** [comment]
@wusyong I didn't do anything like this before and I don't want to mess up even worse. Do you mean pulling from upstream onto my dev branch and then merging with all of _my_ contributions squashed into one commit? Wouldn't this negate all my commit messages?

**Review by amrbashir — APPROVED** [review]
I guess it is ready for merge now, @wusyong would you mind testing and merge?

**Pull request merged**
